packages/instant/src/redux/async_data.ts

```diff
@@ -33,4 +34,40 @@ export const asyncData = {
             store.dispatch(actions.setAvailableAssets([]));
         }
     },
+    fetchAccountInfoAndDispatchToStore: async (store: Store) => {
+        const { providerState } = store.getState();
+        const web3Wrapper = providerState.web3Wrapper;
+        if (providerState.account.state !== AccountState.Loading) {
+            store.dispatch(actions.setAccountStateLoading());
+        }
+        let availableAddresses: string[];
+        try {
+            availableAddresses = await web3Wrapper.getAvailableAddressesAsync();
+        } catch (e) {
+            store.dispatch(actions.setAccountStateError());
+            return;
+        }
+        if (!_.isEmpty(availableAddresses)) {
+            const activeAddress = availableAddresses[0];
+            store.dispatch(actions.setAccountStateReady(activeAddress));
+            await asyncData.fetchAccountBalanceAndDispatchToStore(store);
+        } else {
+            store.dispatch(actions.setAccountStateLocked());
+        }
+    },
+    fetchAccountBalanceAndDispatchToStore: async (store: Store) => {
```

Comment:
Is it possible that we'd have a concurrency issue in which we'd dispatch a balance for an account that they have since changed?   I feel like we may need a similar check here as we have for the buy quote.  

Alternatively, part of me feels like balance fetching could be paired with address fetching.  This could be easier to reason about -- we always get the address and the balance together, or the whole thing is considered "failed".  I'm not married to this though, what do you think?

Reply:
Yes, you are correct wrt to the balance update coming late issue